File: CalCal/Utils/StreakManager.swift
import Foundation
import SwiftData

class StreakManager {
    private let modelContext: ModelContext
    
    init(modelContext: ModelContext) {
        self.modelContext = modelContext
    }
    
    /// Calculates the current streak of consecutive days with at least one food item logged.
    /// A streak continues if there is at least one entry for each consecutive day going backwards from today or yesterday.
    func calculateCurrentStreak() throws -> Int {
        let calendar = Calendar.current
        let now = Date()
        let startOfToday = calendar.startOfDay(for: now)
        
        var streak = 0
        var checkDate = startOfToday
        
        // 1. Check if we have anything today or yesterday to start/continue the streak
        let hasToday = try hasEntries(on: startOfToday)
        let yesterday = calendar.date(byAdding: .day, value: -1, to: startOfToday)!
        let hasYesterday = try hasEntries(on: yesterday)
        
        if !hasToday && !hasYesterday {
            return 0
        }
        
        // If we have nothing today but had something yesterday, the streak is still alive (it's "yesterday's" streak)
        // If we have something today, we start counting from today.
        if !hasToday {
            checkDate = yesterday
        }
        
        // 2. Count backwards
        while true {
            if try hasEntries(on: checkDate) {
                streak += 1
                guard let nextDate = calendar.date(byAdding: .day, value: -1, to: checkDate) else { break }
                checkDate = nextDate
            } else {
                break
            }
        }
        
        return streak
    }
    
    private func hasEntries(on date: Date) throws -> Bool {
        let calendar = Calendar.current
        let startOfDay = calendar.startOfDay(for: date)
        guard let endOfDay = calendar.date(byAdding: .day, value: 1, to: startOfDay) else { return false }
        
        let predicate = #Predicate<FoodItem> { item in
            item.dateEaten >= startOfDay && item.dateEaten < endOfDay
        }
        
        var descriptor = FetchDescriptor<FoodItem>(predicate: predicate)
        descriptor.fetchLimit = 1
        
        let count = try modelContext.fetchCount(descriptor)
        return count > 0
    }
}

File: Android app/app/src/main/java/com/example/portio/domain/util/StreakManager.kt
package com.example.portio.domain.util

import com.example.portio.data.local.FoodItemDao
import java.util.Calendar
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class StreakManager @Inject constructor(private val dao: FoodItemDao) {

    /**
     * Calculates the current streak of consecutive days with at least one food item logged.
     * Streak is alive if there are entries today OR yesterday (mirrors iOS logic).
     */
    suspend fun calculateCurrentStreak(): Int {
        val cal = Calendar.getInstance()
        cal.set(Calendar.HOUR_OF_DAY, 0); cal.set(Calendar.MINUTE, 0)
        cal.set(Calendar.SECOND, 0); cal.set(Calendar.MILLISECOND, 0)
        val startOfToday = cal.timeInMillis

        val hasToday = hasEntries(startOfToday)
        cal.add(Calendar.DAY_OF_MONTH, -1)
        val startOfYesterday = cal.timeInMillis
        val hasYesterday = hasEntries(startOfYesterday)

        if (!hasToday && !hasYesterday) return 0

        var checkDate = if (hasToday) startOfToday else startOfYesterday
        var streak = 0

        while (true) {
            if (hasEntries(checkDate)) {
                streak++
                val prev = Calendar.getInstance().apply {
                    timeInMillis = checkDate
                    add(Calendar.DAY_OF_MONTH, -1)
                }
                checkDate = prev.timeInMillis
            } else {
                break
            }
        }

        return streak
    }

    private suspend fun hasEntries(startOfDay: Long): Boolean {
        val endOfDay = startOfDay + 24 * 60 * 60 * 1000L
        return dao.countForDay(startOfDay, endOfDay) > 0
    }
}
